### ISSUE TITLE
Did you use hard attention using keras in your code before?

### ISSUE THREAD

**Issue by curry111** [fchollet/keras#6244]
If yes, How did you use it in keras?

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by unrealwill** [comment]
Yes but it's not pretty.

I implemented a (currently proprietary) framework on top of Keras to do some reinforcement learning.

I had to write some custom theano op (or you can use pyfunc if you use tensorflow) which will do the sampling and call the non differentiable functor which update the state and return a reward and action likelyhood, then wrap this op into a keras layer. Then another custom layer which collects the rewards, action likelihoods and value functions, and add the generalized advantage calculations necessary to the loss.

It works end to end and converge successfully on easy problems, (it doesn't have replay memory as it process a full episode at a time, and exploration is not yet good, but on the good side I can have hidden cells which keeps a memory between successive actions like a recursive network).
Though I do not recommend doing it keras because it is quite hard to take advantage of multiple machine parallelism and you definitely want this because it will otherwise be quite slow. So you are probably better using a a3s tensorflow implementation and eventually use some keras model for syntactic sugar.

Alternatively you may want to use evolution strategy and just throw enough machines at it to get result a faster but at greater computational cost.
